src/sentry/ratelimits/cardinality.py

```diff
@@ -0,0 +1,139 @@
+
+
+@dataclass(frozen=True)
+class RequestedQuota:
+    # A string that all redis state is prefixed with. For example
+    # `sentry-string-indexer` where 123 is an organization id.
+    prefix: str
+
+    # A unit is an abstract term for the object type we want to limit the
+    # cardinality of.
+    #
+    # For example, if you want to limit the cardinality of timeseries in a
+    # metrics service, this would be a set of hashes composed from `(org_id,
+    # metric_name, tags)`.
+    #
+    # ...though you can probably omit org_id if it is already in the prefix.
+    unit_hashes: Collection[Hash]
+
+    # Which quotas to check against. The number of not-yet-seen hashes must
+    # "fit" into all quotas.
+    quotas: Sequence[Quota]
+
+
+@dataclass(frozen=True)
+class GrantedQuota:
+    # The prefix from RequestedQuota
+    prefix: str
+
+    # The subset of hashes that passed through.
+    granted_unit_hashes: Collection[Hash]
+
+    # If len(granted_unit_hashes) < len(RequestedQuota.unit_hashes), this
+    # contains the quotas that were reached.
+    reached_quotas: Sequence[Quota]
+
+
+class CardinalityLimiter(Service):
+    def __init__(self, **options: Any) -> None:
+        pass
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[Timestamp] = None
+    ) -> Tuple[Timestamp, Sequence[GrantedQuota]]:
+        """
+        Given a set of quotas requests and limits, compute how much quota could
+        be consumed.
+
+        :param requests: The requests to return "grants" for.
+        :param timestamp: The timestamp of the incoming request. Defaults to
+            the current timestamp.
+
+            Providing a too old timestamp here _can_ effectively disable rate
+            limits, as the older request counts may no longer be stored.
+            However, consistently providing old timestamps here will work
+            correctly.
+        """
+        if timestamp is None:
+            timestamp = int(time.time())
+        else:
+            timestamp = int(timestamp)
+
+        grants = [
+            GrantedQuota(
+                prefix=request.prefix, granted_unit_hashes=request.unit_hashes, reached_quotas=[]
+            )
+            for request in requests
+        ]
+
+        return timestamp, grants
+
+    def use_quotas(
+        self,
+        requests: Sequence[RequestedQuota],
+        grants: Sequence[GrantedQuota],
+        timestamp: Timestamp,
+    ) -> None:
+        """
+        Given a set of requests and the corresponding return values from
+        `check_within_quotas`, consume the quotas.
+
+        :param requests: The requests that have previously been passed to
+            `check_within_quotas`.
+        :param timestamp: The request timestamp that has previously been passed
+            to `check_within_quotas`.
+        :param grants: The return value of `check_within_quotas` which
+            indicates how much quota should actually be consumed.
```

Comment:
What about referencing in the Grant class everything we need to call `use_quotas` ? So we do not need to preserve the `RequestedQuota` object and ensure we pass the same sequence and that the Grants are in the same order ?

Reply:
I cargo-culted this from the sliding windows rate limiter as well, but I think you're right